File: src/utils/collectionArchive.ts
import JSZip from "jszip";
import { invokeCommand, isTauri } from "../lib/tauri";
import * as db from "../lib/database";
import { buildAnkiApkg } from "./ankiExport";
import type { Document, Extract, LearningItem } from "../types/document";
import type { Collection } from "../stores/collectionStore";
import type {
  ArchiveFileEntry,
  CollectionArchiveManifest,
  CollectionArchivePayload,
  CollectionExportScope,
  ParsedCollectionArchive,
} from "../types/archive";

function getFilenameFromPath(path: string, fallback: string): string {
  const trimmed = path.split("?")[0];
  const parts = trimmed.split(/[\\/]/);
  return parts[parts.length - 1] || fallback;
}

function base64ToBytes(base64: string): Uint8Array {
  const binary = atob(base64);
  const bytes = new Uint8Array(binary.length);
  for (let i = 0; i < binary.length; i += 1) {
    bytes[i] = binary.charCodeAt(i);
  }
  return bytes;
}

export function collectLocalStorage(): Record<string, string> {
  const data: Record<string, string> = {};
  for (let i = 0; i < localStorage.length; i += 1) {
    const key = localStorage.key(i);
    if (!key) continue;
    const value = localStorage.getItem(key);
    if (value !== null) {
      data[key] = value;
    }
  }
  return data;
}

function buildScopeDocumentSet(
  scope: CollectionExportScope,
  activeCollectionId: string | null,
  documentAssignments: Record<string, string>,
  documents: Document[]
): Set<string> {
  if (scope === "all" || !activeCollectionId) {
    return new Set(documents.map((doc) => doc.id));
  }

  const ids = new Set<string>();
  for (const doc of documents) {
    const assigned = documentAssignments[doc.id];
    if (assigned === activeCollectionId) {
      ids.add(doc.id);
    }
  }
  return ids;
}

export async function buildCollectionArchive(options: {
  scope: CollectionExportScope;
  activeCollectionId: string | null;
  collections: Collection[];
  documentAssignments: Record<string, string>;
}): Promise<{ blob: Blob; filename: string }> {
  const documents = await invokeCommand<Document[]>("get_documents");
  const extracts = await invokeCommand<Extract[]>("get_extracts", {});
  const learningItems = await invokeCommand<LearningItem[]>("get_all_learning_items");

  const scopedDocIds = buildScopeDocumentSet(
    options.scope,
    options.activeCollectionId,
    options.documentAssignments,
    documents
  );

  const scopedDocuments = documents.filter((doc) => scopedDocIds.has(doc.id));
  const scopedExtracts = extracts.filter((extract) => scopedDocIds.has(extract.documentId));
  const scopedExtractIds = new Set(scopedExtracts.map((extract) => extract.id));
  const scopedLearningItems = learningItems.filter((item) => {
    if (item.documentId && scopedDocIds.has(item.documentId)) return true;
    if (item.extractId && scopedExtractIds.has(item.extractId)) return true;
    return false;
  });

  const zip = new JSZip();
  const files: ArchiveFileEntry[] = [];

  for (const doc of scopedDocuments) {
    if (!doc.filePath) continue;
    const base64 = await invokeCommand<string>("read_document_file", { filePath: doc.filePath });
    if (!base64) continue;
    const filename = getFilenameFromPath(doc.filePath, `${doc.id}.bin`);
    const zipPath = `files/${doc.id}/${filename}`;
    const bytes = base64ToBytes(base64);
    zip.file(zipPath, bytes);
    files.push({
      documentId: doc.id,
      filename,
      contentType: doc.fileType ? `application/${doc.fileType}` : undefined,
      zipPath,
      size: bytes.byteLength,
    });
  }

  const deckName = options.scope === "all"
    ? "Incrementum Export"
    : options.collections.find((c) => c.id === options.activeCollectionId)?.name || "Incrementum Export";
  const apkgBytes = await buildAnkiApkg(scopedLearningItems, { deckName });
  zip.file("anki/flashcards.apkg", apkgBytes);

  const settingsRaw = localStorage.getItem("incrementum-settings");
  const settings = settingsRaw ? JSON.parse(settingsRaw) : null;

  const localStorageDump = collectLocalStorage();

  const collectionsPayload =
    options.scope === "all"
      ? {
        collections: options.collections,
        activeCollectionId: options.activeCollectionId,
        documentAssignments: options.documentAssignments,
      }
      : {
        collections: options.collections.filter((c) => c.id === options.activeCollectionId),
        activeCollectionId: options.activeCollectionId,
        documentAssignments: Object.fromEntries(
          Object.entries(options.documentAssignments).filter(([docId]) => scopedDocIds.has(docId))
        ),
      };

  const payload: CollectionArchivePayload = {
    documents: scopedDocuments,
    extracts: scopedExtracts,
    learningItems: scopedLearningItems,
    files,
    collections: collectionsPayload,
    settings,
    localStorage: localStorageDump,
  };

  const manifest: CollectionArchiveManifest = {
    archiveType: "incrementum-collection-export",
    version: "1.0",
    exportedAt: new Date().toISOString(),
    scope: options.scope,
  };

  zip.file("manifest.json", JSON.stringify(manifest, null, 2));
  zip.file("data/payload.json", JSON.stringify(payload, null, 2));

  const blob = await zip.generateAsync({ type: "blob" });
  const date = new Date().toISOString().split("T")[0];
  const scopeLabel = options.scope === "all" ? "all" : "collection";
  const filename = `incrementum-${scopeLabel}-export-${date}.zip`;

  return { blob, filename };
}

export async function parseCollectionArchive(file: File): Promise<ParsedCollectionArchive> {
  const buffer = await file.arrayBuffer();
  const zip = await JSZip.loadAsync(buffer);
  const manifestRaw = await zip.file("manifest.json")?.async("string");
  if (!manifestRaw) {
    throw new Error("Archive manifest not found.");
  }
  const manifest = JSON.parse(manifestRaw) as CollectionArchiveManifest;
  if (manifest.archiveType !== "incrementum-collection-export") {
    throw new Error("Unsupported archive type.");
  }
  const payloadRaw = await zip.file("data/payload.json")?.async("string");
  if (!payloadRaw) {
    throw new Error("Archive payload missing.");
  }
  const payload = JSON.parse(payloadRaw) as CollectionArchivePayload;

  const files: ParsedCollectionArchive["files"] = [];
  for (const entry of payload.files) {
    const fileBlob = await zip.file(entry.zipPath)?.async("blob");
    if (!fileBlob) continue;
    files.push({ entry, blob: fileBlob });
  }

  return { manifest, payload, files };
}

export function restoreLocalStorage(data: Record<string, string>): void {
  localStorage.clear();
  Object.entries(data).forEach(([key, value]) => {
    localStorage.setItem(key, value);
  });
}

export async function restoreBrowserArchive(parsed: ParsedCollectionArchive): Promise<void> {
  const filesToStore: db.StoredFile[] = [];
  const filePathByDocument: Record<string, string> = {};

  for (const { entry, blob } of parsed.files) {
    const filePath = `archive-file://${entry.documentId}/${entry.filename}`;
    const file = new File([blob], entry.filename, {
      type: entry.contentType || "application/octet-stream",
    });
    filesToStore.push({
      id: filePath,
      filename: entry.filename,
      content_type: entry.contentType || file.type,
      blob: file,
      created_at: new Date().toISOString(),
    });
    filePathByDocument[entry.documentId] = filePath;
  }

  const documents = (parsed.payload.documents as Document[]).map((doc) => {
    const replacementPath = filePathByDocument[doc.id];
    if (!replacementPath) return doc;
    return { ...doc, filePath: replacementPath };
  });

  const extracts = parsed.payload.extracts as Extract[];
  const learningItems = parsed.payload.learningItems as LearningItem[];

  await db.clearStore("documents");
  await db.clearStore("extracts");
  await db.clearStore("learning_items");
  await db.clearStore("files");

  await db.bulkPutDocuments(documents.map((doc) => ({
    id: doc.id,
    title: doc.title,
    file_path: doc.filePath,
    file_type: doc.fileType,
    content: doc.content,
    content_hash: doc.contentHash,
    total_pages: doc.totalPages,
    current_page: doc.currentPage || 1,
    current_scroll_percent: doc.currentScrollPercent,
    current_cfi: doc.currentCfi,
    current_view_state: doc.currentViewState,
    position_json: doc.positionJson,
    progress_percent: doc.progressPercent,
    category: doc.category,
    tags: doc.tags || [],
    date_added: doc.dateAdded,
    date_modified: doc.dateModified,
    date_last_reviewed: doc.dateLastReviewed,
    extract_count: doc.extractCount,
    learning_item_count: doc.learningItemCount,
    priority_rating: doc.priorityRating,
    priority_slider: doc.prioritySlider,
    priority_score: doc.priorityScore,
    is_archived: doc.isArchived,
    is_favorite: doc.isFavorite,
    metadata: doc.metadata,
    cover_image_url: doc.coverImageUrl,
    cover_image_source: doc.coverImageSource,
    next_reading_date: doc.nextReadingDate,
    reading_count: doc.readingCount,
    stability: doc.stability,
    difficulty: doc.difficulty,
    reps: doc.reps,
    total_time_spent: doc.totalTimeSpent,
  } as db.Document)));

  await db.bulkPutExtracts(extracts.map((extract) => ({
    id: extract.id,
    document_id: extract.documentId,
    content: extract.content,
    html_content: (extract as Extract & { htmlContent?: string }).htmlContent,
    source_url: (extract as Extract & { sourceUrl?: string }).sourceUrl,
    page_title: extract.pageTitle,
    page_number: extract.pageNumber,
    selection_context: extract.selectionContext,
    highlight_color: extract.highlightColor,
    notes: extract.notes,
    progressive_disclosure_level: extract.progressiveDisclosureLevel,
    max_disclosure_level: extract.maxDisclosureLevel,
    date_created: extract.dateCreated,
    date_modified: extract.dateModified,
    tags: extract.tags || [],
    category: extract.category,
    memory_state: (extract as any).memoryState,
    next_review_date: (extract as any).nextReviewDate,
    last_review_date: (extract as any).lastReviewDate,
    review_count: (extract as any).reviewCount ?? 0,
    reps: (extract as any).reps ?? 0,
  } as db.Extract)));

  await db.bulkPutLearningItems(learningItems.map((item) => ({
    id: item.id,
    extract_id: item.extractId,
    document_id: item.documentId,
    item_type: item.itemType,
    question: item.question,
    answer: item.answer,
    cloze_text: item.clozeText,
    cloze_ranges: item.clozeRanges,
    difficulty: item.difficulty,
    interval: item.interval,
    ease_factor: item.easeFactor,
    due_date: item.dueDate,
    date_created: item.dateCreated,
    date_modified: item.dateModified,
    last_review_date: item.lastReviewDate,
    review_count: item.reviewCount,
    lapses: item.lapses,
    state: item.state,
    is_suspended: item.isSuspended,
    tags: item.tags || [],
    memory_state: (item as any).memoryState,
  } as db.LearningItem)));

  if (filesToStore.length > 0) {
    await db.bulkPutFiles(filesToStore);
  }
}

export function shouldUseTauriImport(file: File): boolean {
  const filePath = (file as File & { path?: string }).path;
  return isTauri() && Boolean(filePath);
}

File: src-tauri/src/commands/collection_archive.rs
//! Collection export/import archive commands

use std::fs;
use std::io::Read;
use std::path::PathBuf;

use chrono::Utc;
use serde::Deserialize;
use tauri::{AppHandle, Manager, State};
use zip::ZipArchive;

use crate::database::Repository;
use crate::error::{AppError, Result};
use crate::models::{Document, Extract, LearningItem};

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct ArchiveManifest {
    archive_type: String,
    version: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct ArchiveFileEntry {
    document_id: String,
    filename: String,
    content_type: Option<String>,
    zip_path: String,
}

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct ArchivePayload {
    documents: Vec<Document>,
    extracts: Vec<Extract>,
    learning_items: Vec<LearningItem>,
    files: Vec<ArchiveFileEntry>,
}

fn normalize_file_type(file_type: &crate::models::FileType) -> String {
    format!("{:?}", file_type).to_lowercase()
}

fn normalize_item_type(item_type: &crate::models::ItemType) -> String {
    format!("{:?}", item_type).to_lowercase()
}

fn normalize_item_state(state: &crate::models::ItemState) -> String {
    format!("{:?}", state).to_lowercase()
}

fn ensure_document_storage(app: &AppHandle) -> Result<PathBuf> {
    let app_dir = app
        .path()
        .app_data_dir()
        .map_err(|e| AppError::Internal(format!("Failed to get app data dir: {}", e)))?;
    let docs_dir = app_dir.join("documents");
    fs::create_dir_all(&docs_dir)
        .map_err(|e| AppError::Internal(format!("Failed to create documents dir: {}", e)))?;
    Ok(docs_dir)
}

fn read_zip_bytes(archive: &mut ZipArchive<fs::File>, path: &str) -> Result<Vec<u8>> {
    let mut file = archive
        .by_name(path)
        .map_err(|e| AppError::Internal(format!("Missing archive entry '{}': {}", path, e)))?;
    let mut buffer = Vec::new();
    file.read_to_end(&mut buffer)
        .map_err(|e| AppError::Internal(format!("Failed to read archive entry '{}': {}", path, e)))?;
    Ok(buffer)
}

#[tauri::command]
pub async fn import_collection_archive(
    archive_path: String,
    repo: State<'_, Repository>,
    app: AppHandle,
) -> Result<String> {
    let archive_file = fs::File::open(&archive_path)
        .map_err(|e| AppError::Internal(format!("Failed to open archive: {}", e)))?;
    let mut archive = ZipArchive::new(archive_file)
        .map_err(|e| AppError::Internal(format!("Invalid archive: {}", e)))?;

    let manifest_bytes = read_zip_bytes(&mut archive, "manifest.json")?;
    let manifest: ArchiveManifest = serde_json::from_slice(&manifest_bytes)
        .map_err(|e| AppError::Internal(format!("Invalid manifest: {}", e)))?;
    if manifest.archive_type != "incrementum-collection-export" {
        return Err(AppError::Internal("Unsupported archive type".to_string()));
    }
    if manifest.version != "1.0" {
        return Err(AppError::Internal("Unsupported archive version".to_string()));
    }

    let payload_bytes = read_zip_bytes(&mut archive, "data/payload.json")?;
    let mut payload: ArchivePayload = serde_json::from_slice(&payload_bytes)
        .map_err(|e| AppError::Internal(format!("Invalid archive payload: {}", e)))?;

    let docs_dir = ensure_document_storage(&app)?;

    for file_entry in &payload.files {
        let file_bytes = read_zip_bytes(&mut archive, &file_entry.zip_path)?;
        let doc_dir = docs_dir.join(&file_entry.document_id);
        fs::create_dir_all(&doc_dir)
            .map_err(|e| AppError::Internal(format!("Failed to create doc dir: {}", e)))?;
        let file_path = doc_dir.join(&file_entry.filename);
        fs::write(&file_path, file_bytes)
            .map_err(|e| AppError::Internal(format!("Failed to write document file: {}", e)))?;

        if let Some(doc) = payload.documents.iter_mut().find(|d| d.id == file_entry.document_id) {
            doc.file_path = file_path.to_string_lossy().to_string();
        }
    }

    let pool = repo.pool();
    let mut tx = pool.begin().await?;

    sqlx::query("DELETE FROM learning_items").execute(&mut *tx).await?;
    sqlx::query("DELETE FROM extracts").execute(&mut *tx).await?;
    sqlx::query("DELETE FROM documents").execute(&mut *tx).await?;

    for doc in &payload.documents {
        let tags_json = serde_json::to_string(&doc.tags)?;
        let metadata_json = doc.metadata.as_ref().map(serde_json::to_string).transpose()?;
        let file_type_str = normalize_file_type(&doc.file_type);

        sqlx::query(
            r#"
            INSERT INTO documents (
                id, title, file_path, file_type, content, content_hash,
                total_pages, current_page, current_scroll_percent, current_cfi, current_view_state,
                position_json, progress_percent, category, tags, date_added, date_modified, date_last_reviewed,
                extract_count, learning_item_count, priority_rating, priority_slider, priority_score,
                is_archived, is_favorite, metadata, cover_image_url, cover_image_source,
                next_reading_date, reading_count, stability, difficulty, reps, total_time_spent, consecutive_count
            ) VALUES (
                ?1, ?2, ?3, ?4, ?5, ?6,
                ?7, ?8, ?9, ?10, ?11,
                ?12, ?13, ?14, ?15, ?16, ?17, ?18,
                ?19, ?20, ?21, ?22, ?23,
                ?24, ?25, ?26, ?27, ?28,
                ?29, ?30, ?31, ?32, ?33, ?34, ?35
            )
            "#,
        )
        .bind(&doc.id)
        .bind(&doc.title)
        .bind(&doc.file_path)
        .bind(&file_type_str)
        .bind(&doc.content)
        .bind(&doc.content_hash)
        .bind(doc.total_pages)
        .bind(doc.current_page)
        .bind(doc.current_scroll_percent)
        .bind(&doc.current_cfi)
        .bind(&doc.current_view_state)
        .bind(&doc.position_json)
        .bind(doc.progress_percent)
        .bind(&doc.category)
        .bind(&tags_json)
        .bind(doc.date_added)
        .bind(doc.date_modified)
        .bind(doc.date_last_reviewed)
        .bind(doc.extract_count)
        .bind(doc.learning_item_count)
        .bind(doc.priority_rating)
        .bind(doc.priority_slider)
        .bind(doc.priority_score)
        .bind(doc.is_archived)
        .bind(doc.is_favorite)
        .bind(metadata_json)
        .bind(&doc.cover_image_url)
        .bind(&doc.cover_image_source)
        .bind(doc.next_reading_date)
        .bind(doc.reading_count)
        .bind(doc.stability)
        .bind(doc.difficulty)
        .bind(doc.reps)
        .bind(doc.total_time_spent)
        .bind(doc.consecutive_count)
        .execute(&mut *tx)
        .await?;
    }

    for extract in &payload.extracts {
        let tags_json = serde_json::to_string(&extract.tags)?;
        let selection_context_json = extract
            .selection_context
            .as_ref()
            .map(serde_json::to_string)
            .transpose()?;

        let (stability, difficulty) = extract
            .memory_state
            .as_ref()
            .map(|s| (Some(s.stability), Some(s.difficulty)))
            .unwrap_or((None, None));

        sqlx::query(
            r#"
            INSERT INTO extracts (
                id, document_id, content, html_content, source_url, page_title, page_number,
                selection_context, highlight_color, notes, progressive_disclosure_level,
                max_disclosure_level, date_created, date_modified, tags, category,
                memory_state_stability, memory_state_difficulty, next_review_date, last_review_date,
                review_count, reps
            ) VALUES (
                ?1, ?2, ?3, ?4, ?5, ?6, ?7,
                ?8, ?9, ?10, ?11,
                ?12, ?13, ?14, ?15, ?16,
                ?17, ?18, ?19, ?20,
                ?21, ?22
            )
            "#,
        )
        .bind(&extract.id)
        .bind(&extract.document_id)
        .bind(&extract.content)
        .bind(&extract.html_content)
        .bind(&extract.source_url)
        .bind(&extract.page_title)
        .bind(extract.page_number)
        .bind(selection_context_json)
        .bind(&extract.highlight_color)
        .bind(&extract.notes)
        .bind(extract.progressive_disclosure_level)
        .bind(extract.max_disclosure_level)
        .bind(extract.date_created)
        .bind(extract.date_modified)
        .bind(&tags_json)
        .bind(&extract.category)
        .bind(stability)
        .bind(difficulty)
        .bind(extract.next_review_date)
        .bind(extract.last_review_date)
        .bind(extract.review_count)
        .bind(extract.reps)
        .execute(&mut *tx)
        .await?;
    }

    for item in &payload.learning_items {
        let tags_json = serde_json::to_string(&item.tags)?;
        let cloze_ranges_json = item
            .cloze_ranges
            .as_ref()
            .map(serde_json::to_string)
            .transpose()?;
        let (stability, difficulty) = item
            .memory_state
            .as_ref()
            .map(|s| (Some(s.stability), Some(s.difficulty)))
            .unwrap_or((None, None));

        let item_type_str = normalize_item_type(&item.item_type);
        let state_str = normalize_item_state(&item.state);

        sqlx::query(
            r#"
            INSERT INTO learning_items (
                id, extract_id, document_id, item_type, question,
                answer, cloze_text, cloze_ranges, difficulty, interval,
                ease_factor, due_date, date_created, date_modified,
                last_review_date, review_count, lapses, state,
                is_suspended, tags, memory_state_stability, memory_state_difficulty
            ) VALUES (
                ?1, ?2, ?3, ?4, ?5,
                ?6, ?7, ?8, ?9, ?10,
                ?11, ?12, ?13, ?14,
                ?15, ?16, ?17, ?18,
                ?19, ?20, ?21, ?22
            )
            "#,
        )
        .bind(&item.id)
        .bind(&item.extract_id)
        .bind(&item.document_id)
        .bind(&item_type_str)
        .bind(&item.question)
        .bind(&item.answer)
        .bind(&item.cloze_text)
        .bind(cloze_ranges_json)
        .bind(item.difficulty)
        .bind(item.interval)
        .bind(item.ease_factor)
        .bind(item.due_date)
        .bind(item.date_created)
        .bind(item.date_modified)
        .bind(item.last_review_date)
        .bind(item.review_count)
        .bind(item.lapses)
        .bind(&state_str)
        .bind(item.is_suspended)
        .bind(&tags_json)
        .bind(stability)
        .bind(difficulty)
        .execute(&mut *tx)
        .await?;
    }

    tx.commit().await?;

    Ok(format!(
        "Imported {} documents, {} extracts, {} learning items at {}",
        payload.documents.len(),
        payload.extracts.len(),
        payload.learning_items.len(),
        Utc::now().to_rfc3339()
    ))
}
